File: Part 1/Chapter 6 - Simple Algebraic Data Types/jmcarter_solutions.cpp
// Question 2 - 3 - 4


struct Shape {
    virtual double area() = 0
    virtual double circ() = 0  // new, Q.3
}

struct Circle : public Shape {
    Circle(double radius) : r{radius} {}
    double area() override {
        return PI*r*r;
    }
    double circ() override {   // new, Q.3
        return 2*PI*r;
    }
private:
    double r;
}

struct Rect : public Shape {
    Rect(double height, double width) : h{height}, w{width} {}
    double area() override {
        return h*w;
    }
    double circ() override {    // new Q.3
        return 2*(h+w);
    }
private:
    double h;
    double w;
}

// New Q.4//
struct Square : public Shape {
    Rect(double side) : s{side} {}
    double area() override {
        return s*s;
    }
    double circ() override {
        return 4*s;
    }
private:
    double s;
}

File: Part 1/Chapter 6 - Simple Algebraic Data Types/jmcarter_solutions.hs
-- Question 1

maybe_to_either :: Maybe a -> Either () a
maybe_to_either x =
    case x of
        Nothing -> Left ()
        Just a -> Right a

either_to_maybe :: Either () a -> Maybe a
either_to_maybe x =
    case x of
        Left () -> Nothing
        Right a -> Just a


-- Questions 4. Add new shape to Hierarchy

data Shape = Circle Float
            | Rect Float Float
            | Square Float  -- new

area :: Shape -> Float
area (Circle r) = pi * r * r
area (Rect d h) = d * h
area (Square s) = s * s     -- new

circ :: Shape -> Float
circ (Circle r) = 2.0 * pi * r
circ (Rect d h) = 2.0 * (d + h)
circ (Square s) = 4.0 * s   -- new


-- Question 5.

either_to_pair :: Either a a -> (a, Bool)
either_to_pair (Left a)  = (a, True)
either_to_pair (Right a) = (a, False)

pair_to_either :: (a, Bool) -> Either a a
pair_to_either (a, True) = Left a
pair_to_either (a, False)  = Right a


data Element = Element{ name :: String, symbol :: String, atomicNumber :: Int }

